### ISSUE TITLE
Enhancement: Add option to provide Airflow builtins

### ISSUE THREAD

**Issue by dmohns** [sqlfluff/sqlfluff#3242]
### Search before asking

- [X] I searched the [issues](https://github.com/sqlfluff/sqlfluff/issues) and found no similar issues.

### Description

Airflow is commonly used to run and schedule SQL queries on various databases. It's even referenced in SQLFluff documentation https://docs.sqlfluff.com/en/stable/realworld.html#modularity

Currently, all users that want to integrate SQLFLuff into their Airflow code base have to go through the same process of understanding, how to

- Add custom Variables
- Add custom Macros
- Add custom Filters

into SQLFluff before they can start using SQLFluff. With the latter (custom Filters) currently not even possible.

It would be great if SQLFluff had first class support for Airflow. One idea would be to provide an option `apply_airflow_builtins` that acts similar as `apply_dbt_builtins`, but provides variables, macros and filters as per https://airflow.apache.org/docs/apache-airflow/stable/templates-ref.html

### Use case

Greatly increase the user experience of Airflow users, by reducing the number of customisation needed to rollout SQLFluff to Airflow code.

### Dialect

all

### Are you willing to work on and submit a PR to address the issue?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/sqlfluff/sqlfluff/blob/main/CODE_OF_CONDUCT.md)

**Comment by barrywhart** [comment]
@dmohns: Have you already defined these in your configuration? Care to share that? That would make it easier to add this.

**Comment by dmohns** [comment]
I don't use `configuration` but a library to implement the Airflow variables and macros. This is because Airflow provided macros need to be in `macro` name space. And also to be able to more complex templating like `{{ task_instance.logical_date }}`.

What I have is pretty similar to the code in https://github.com/sqlfluff/sqlfluff/pull/3245 if you want to have a first glance.

**Comment by dmohns** [comment]
Some preliminary work has been done on https://github.com/sqlfluff/sqlfluff/pull/3245

If someone wants to take this, they can take the PR as a starting point or source of ideas how to tackle this.

**Comment by tdstark** [comment]
@dmohns I'm thinking about taking this. Any chance you could get me up to speed on what's left over?

**Comment by dmohns** [comment]
Hey @tdstark yes, sure! I would love to see this getting picked up.

You can take the PR https://github.com/sqlfluff/sqlfluff/pull/3245 as a starting point. Feel free to fork it, I'm not sure how much this SQLFluff code base has changed since then, probably it will need some refactoring/rebase.

In general, I was trying to write the `airflow_builtins` as modular as possible, using SQLFluff libraries codes which posed some challenges, see below.

The following features I was trying to implement, amongst the to-dos

### Airflow variables

Here I added a couple of Emulators to handle most common Airflow variable usages. For example here: https://github.com/sqlfluff/sqlfluff/pull/3245/files#diff-bf17d9c66c252e1e6d0d9d03478cf9baae80de1cdf839622c16bf84e229a395cR74

**To Do:**

- Fix the coverage tests. They are complaining about certain lines not covered, but they are pretty clearly used in the tests. I couldn't figure out what was the problem.
- Enrich the Tasks and DAG emulator objects. I only added their basic properties to test the feasibility of this. It would be nice if we could get more a less the entire Airflow Python API available.

### Airflow filters

All Airflow filters are implemented and work quite well. However the way these get injected into SQLFluff's Jinja context feels extremely hacky, see https://github.com/sqlfluff/sqlfluff/pull/3245/files#diff-508f402c74d53f489e25725bc3a04facf1697bb58efdaa7114acc80fc2a8609bR285-R286

There is currently no easy way to add Jinja filters from libraries, see this issue also: https://github.com/sqlfluff/sqlfluff/issues/3074

### Airflow Macros

All Airflow macros are implemented.

**To Do**

- Fix Coverage tests. Similar to above, I have no idea why the tests where complaining here and how to fix them.

### Airflow XCom 

This was entirely work-in-progress and I couldn't get it done. I wanted to achieve a way for users to provide an easy way to XCom pull values, for example like so

```
[sqlfluff:templater:jinja]
apply_airflow_builtins = True

[sqlfluff:templater:jinja_builtins:airflow:xcom_pull]
task_1 = {"return_value": "some_return_value"}
task_2 = {
    "return_value": "some_return_value",
    "xcom_value_int": 1,
    "xcom_value_list": ["a", "b", "c"]
    }
```

for

```
{{ ti.xcom_pull(task_ids=task_2, key=return_value) }}
```

However I couldn't find a way to access the SQFluff config from the library.

**To Do**

- Either implement the feature as described above, or skip. In case of the latter, maybe we can create a new Github ticket to track addition of this feature.

### Misc

**To Do**

- From what I remember there were some issues with MyPy. These need to be solved eventually, assuming they still appear after rebasing.
- Add the (user facing) documentation explaining how to use `airflow_builtins` amongst some examples.

**Comment by tdstark** [comment]
This is helpful. Thank you @dmohns. Was any thought put into turning this into a plugin? I have a lingering sense that since this work is Airflow specific, it should probably be a plugin. I'm not sure if the architecture would support this or not though, since I'm not sure if we're using plugins in any specific way. @barrywhart @tunetheweb any thoughts?

**Comment by dmohns** [comment]
> This is helpful. Thank you @dmohns. Was any thought put into turning this into a plugin? I have a lingering sense that since this work is Airflow specific, it should probably be a plugin. I'm not sure if the architecture would support this or not though, since I'm not sure if we're using plugins in any specific way. @barrywhart @tunetheweb any thoughts?

Not sure either. If I recall correctly there was some brief mentioning of the idea of turning it into a plugin. I have never really explored the capabilities of SQLFluff plugins. But looking at specifically the hacks required to get filters to work, I doubt it could be fully achieved as a plugin. But it would be nice to get some insight from someone more familiar with the plugin API 😉

**Comment by barrywhart** [comment]
Currently, SQLFluff only supports templaters and rules as plugins.

In principle, this sounds reasonable, and SQLFluff probably shouldn't have a large number of extra macros provided as builtins. However, do note that SQLFluff already provides some basic dbt macros as builtins for the jinja templater (documented [here](https://docs.sqlfluff.com/en/stable/configuration.html#builtin-macro-blocks)). Could these Airflow macros work similarly?

I think it would be okay to add the Airflow macros as builtins, then we could later add support for macro plugins and split them out. This is essentially what we did with the dbt templater -- it was part of the core SQLFluff package until about a year ago.

**Comment by dmohns** [comment]
> In principle, this sounds reasonable, and SQLFluff probably shouldn't have a large number of extra macros provided as builtins. However, do note that SQLFluff already provides some basic dbt macros as builtins for the jinja templater (documented [here](https://docs.sqlfluff.com/en/stable/configuration.html#builtin-macro-blocks)). Could these Airflow macros work similarly?

Yes, exactly. The initial work done was following the old DBT implementation closely.

> I think it would be okay to add the Airflow macros as builtins, then we could later add support for macro plugins and split them out. This is essentially what we did with the dbt templater -- it was part of the core SQLFluff package until about a year ago.

This was also my idea. For now, have this builtin due to the limitations mentioned above, but trying to write the code as modular as possible. Then once the plugin system is capable enough, it would be an easy task to move this out.
